File: src/athena/utils/single_turn_chat.py
"""One-off Agent interaction with optional caller-owned conversation memory."""

import asyncio
from typing import TYPE_CHECKING, Any
from uuid import uuid4

from pydantic_ai.messages import ModelMessage, ModelResponse

from athena.core.agent import AgentContext, create_agent
from athena.core.thread_models import AthenaThread, AthenaTurn
from athena.core.tool import ToolRegistry
from athena.core.tool_types import EmitEvent
from athena.memory.context_manager import ContextManager

if TYPE_CHECKING:
    from openai import AsyncOpenAI


async def _noop_emit(
    _kind: str, _ref: str, _data: dict[str, Any] | None = None
) -> None:
    """丢弃所有事件。"""


def _validate(
    prompt: str,
    model: str,
    max_turns: int,
    max_tokens: int,
    temperature: float,
) -> None:
    if not isinstance(prompt, str) or not prompt.strip():
        raise ValueError("prompt must be a non-empty string")
    if not isinstance(model, str) or not model.strip():
        raise ValueError("model must be a non-empty string")
    if isinstance(max_turns, bool) or not isinstance(max_turns, int) or max_turns <= 0:
        raise ValueError("max_turns must be a positive integer")
    if (
        isinstance(max_tokens, bool)
        or not isinstance(max_tokens, int)
        or max_tokens <= 0
    ):
        raise ValueError("max_tokens must be a positive integer")
    if (
        isinstance(temperature, bool)
        or not isinstance(temperature, (int, float))
        or not 0 <= temperature <= 2
    ):
        raise ValueError("temperature must be between 0 and 2")


def _final_text(messages: list[ModelMessage]) -> str:
    for message in reversed(messages):
        if not isinstance(message, ModelResponse):
            continue
        if any(part.part_kind == "tool-call" for part in message.parts):
            continue
        text = "".join(
            part.content
            for part in message.parts
            if part.part_kind == "text" and isinstance(part.content, str)
        )
        if text.strip():
            return text
    raise RuntimeError("single-turn chat produced no final assistant text")


async def single_turn_chat(
    prompt: str,
    *,
    model: str,
    memory: ContextManager | None = None,
    tools: ToolRegistry | None = None,
    system_prompt: str | None = None,
    client: "AsyncOpenAI | None" = None,
    max_turns: int = 20,
    max_tokens: int = 4096,
    temperature: float = 0.1,
    emit: EmitEvent | None = None,
    cancel: asyncio.Event | None = None,
) -> str:
    """Run one Agent turn and return its final text.

    Omitting ``memory`` makes the call stateless. Supplying a ``ContextManager``
    lets the caller retain this turn and reuse it in later calls. Callers must not
    use the same mutable memory concurrently.
    """

    _validate(prompt, model, max_turns, max_tokens, temperature)
    active_cancel = cancel if cancel is not None else asyncio.Event()
    if active_cancel.is_set():
        raise asyncio.CancelledError

    active_memory = memory if memory is not None else ContextManager()
    active_tools = tools if tools is not None else ToolRegistry()
    start_index = active_memory.snapshot()[0]
    identity = uuid4().hex
    thread_id = f"thread:single-turn:{identity}"
    turn_id = f"turn:single-turn:{identity}"
    thread = AthenaThread(
        thread_id=thread_id,
        session_id=f"session:single-turn:{identity}",
        status="running",
        context_ref=f"context:single-turn:{identity}",
    )
    turn = AthenaTurn(
        turn_id=turn_id,
        thread_id=thread_id,
        request_ref=prompt,
        status="running",
    )
    agent = create_agent(
        model=model,
        tools=active_tools,
        system_prompt=system_prompt or "",
        client=client,
        max_turns=max_turns,
        max_tokens=max_tokens,
        temperature=temperature,
        name="single-turn-chat",
    )
    context = AgentContext(
        thread=thread,
        turn=turn,
        emit=emit if emit is not None else _noop_emit,
        tools=active_tools,
        cancel=active_cancel,
        memory=active_memory,
        input_text=prompt,
    )

    await agent.run(context)

    return _final_text(active_memory.items[start_index:])

File: athena_ts/packages/athena-agent/src/single-turn-chat.ts
/**
 * One-off Agent interaction with optional caller-owned conversation memory（移植 utils/single_turn_chat.py）。
 */

import { randomBytes } from "node:crypto"

import { AthenaThreadSchema, AthenaTurnSchema } from "@athena/core"

import { CancelledError } from "./agent/types.js"
import { AgentContext } from "./agent/models.js"
import { createAgent } from "./agent/runtime.js"
import type { ChatClient } from "./agent/settings.js"
import { ToolRegistry } from "./tool.js"
import type { EmitEvent } from "./tool-types.js"
import { ContextManager } from "./memory/context-manager.js"
import type { ModelMessage } from "./messages.js"

async function noopEmit(
  _kind: string,
  _ref: string,
  _data?: Record<string, unknown> | null
): Promise<void> {
  void 0
}

function validate(
  prompt: string,
  model: string,
  maxTurns: number,
  maxTokens: number,
  temperature: number
): void {
  if (typeof prompt !== "string" || !prompt.trim()) {
    throw new Error("prompt must be a non-empty string")
  }
  if (typeof model !== "string" || !model.trim()) {
    throw new Error("model must be a non-empty string")
  }
  if (typeof maxTurns !== "number" || !Number.isInteger(maxTurns) || maxTurns <= 0) {
    throw new Error("max_turns must be a positive integer")
  }
  if (typeof maxTokens !== "number" || !Number.isInteger(maxTokens) || maxTokens <= 0) {
    throw new Error("max_tokens must be a positive integer")
  }
  if (typeof temperature !== "number" || !(temperature >= 0 && temperature <= 2)) {
    throw new Error("temperature must be between 0 and 2")
  }
}

function finalText(messages: ModelMessage[]): string {
  for (let i = messages.length - 1; i >= 0; i--) {
    const message = messages[i]!
    if (message.kind !== "response") continue
    if (message.parts.some((p) => p.part_kind === "tool-call")) continue
    const text = message.parts
      .filter((p) => p.part_kind === "text" && "content" in p && typeof p.content === "string")
      .map((p) => ("content" in p ? p.content : ""))
      .join("")
    if (text.trim()) return text
  }
  throw new Error("single-turn chat produced no final assistant text")
}

export interface SingleTurnChatOptions {
  model: string
  memory?: ContextManager | null
  tools?: ToolRegistry | null
  systemPrompt?: string | null
  client?: ChatClient | null
  maxTurns?: number
  maxTokens?: number
  temperature?: number
  emit?: EmitEvent | null
  cancel?: AbortSignal | null
}

export async function singleTurnChat(prompt: string, opts: SingleTurnChatOptions): Promise<string> {
  const {
    model,
    memory = null,
    tools = null,
    systemPrompt = null,
    client = null,
    maxTurns = 200,
    maxTokens = 4096,
    temperature = 0.1,
    emit = null,
    cancel = null,
  } = opts

  validate(prompt, model, maxTurns, maxTokens, temperature)
  const activeCancel = cancel ?? new AbortController().signal
  if (activeCancel.aborted) throw new CancelledError()

  const activeMemory = memory ?? new ContextManager()
  const activeTools = tools ?? new ToolRegistry()
  const startIndex = activeMemory.snapshot()[0]
  const identity = randomBytes(16).toString("hex")
  const threadId = `thread:single-turn:${identity}`
  const turnId = `turn:single-turn:${identity}`

  const thread = AthenaThreadSchema.parse({
    thread_id: threadId,
    session_id: `session:single-turn:${identity}`,
    status: "running",
    context_ref: `context:single-turn:${identity}`,
  })
  const turn = AthenaTurnSchema.parse({
    turn_id: turnId,
    thread_id: threadId,
    request_ref: prompt,
    status: "running",
  })

  const agent = createAgent(model, activeTools, systemPrompt ?? "", {
    client,
    maxTurns,
    maxTokens,
    temperature,
    name: "single-turn-chat",
  })
  const context = new AgentContext(
    thread,
    turn,
    emit ?? noopEmit,
    activeTools,
    activeCancel,
    activeMemory,
    prompt
  )

  await agent.run(context)
  return finalText(activeMemory.items.slice(startIndex))
}
